pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -83,10 +89,43 @@ func (s *preFilterState) Clone() framework.StateData {
 // failed by this plugin schedulable.
 func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 	}
 }
 
+// isSchedulableAfterNodeChange is invoked whenever a node changed. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *NodeAffinity) isSchedulableAfterNodeChange(pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		pl.logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	requiredNodeAffinity := nodeaffinity.GetRequiredNodeAffinity(pod)
+	wasMatched := false
+	if originalNode != nil {
+		wasMatched, err = requiredNodeAffinity.Match(originalNode)
+		if err != nil {
+			pl.logger.Error(err, "failed to match pod to the original node", "pod", klog.KObj(pod), "node", klog.KObj(originalNode))
+			return framework.QueueAfterBackoff
+		}
+	}
+	isMatched, err := requiredNodeAffinity.Match(modifiedNode)
+	if err != nil {
+		pl.logger.Error(err, "failed to match pod to the modified node", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueAfterBackoff
+	}
+	if !wasMatched && isMatched {
+		// This modification makes this Node match with Pod's NodeAffinity.
+		pl.logger.V(4).Info("node was created or updated, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueAfterBackoff
```

Comment:
Well.. we were talking about the QueueImmediate vs QueueAfterBackoff here
https://github.com/kubernetes/kubernetes/pull/117561#discussion_r1186573820

In this case, given the fact that a Pod may be unschedulable again even if the Node is changed to match with the NodeAffinity, I think we should return QueueAfterBackoff.

Reply:
The pod may always be unschedulable, even if we return immediate. The point is to have a reasonable guess that something substantial changed to make the pod schedulable. 

What if all previously failed plugins returned `QueueAfterBackoff` we translate that to `QueueImmediate`?

Reply:
That means that we eliminate backoffQ, right? It's similar to my past question:

> (Why not we don't put Pods into activeQ soon when events, which may make Pod schedulable, happen?)
https://github.com/kubernetes/kubernetes/pull/117561

Answers (from [Aldo](https://github.com/kubernetes/kubernetes/pull/117561#discussion_r1189061873) and [Wei](https://github.com/kubernetes/kubernetes/pull/117561#discussion_r1189214800)) made sense to me -> if we remove backoffQ completely, high-priority Pods, which fails many time, may keep blocking other lower-priority Pods.

Reply:
We are not completely removing the backoff because this is not suggesting to remove the backoff in all cases, it is only in the case where the cluster state changes to the degree where all previously failed plugins could pass. 

In any case, we can evaluate this later.

Reply:
FMPOV, the decisions made by each plugin should be independent. In this case `node_affinity` can make sure this event will let the pod pass it's `filter`, so it should return the `QueueImmediate`. Of course, this pod may still fail other plugin's filter and be unschedulable again. But I don't think this should be considered by the plugin itself. Maybe we can change the logic in `isPodWorthRequeuing` to coordinate this.  Like only requeue this pod immediately if the majority of plugins responded with  `QueueImmediate` or something else.

Reply:
Also think we should return immediately here, but before the conclusion, I have a question here https://github.com/kubernetes/kubernetes/pull/118551#discussion_r1267676030

Reply:
I have questions to you.
So, if we return QueueImmediate here in NodeAffinity, when to use QueueAfterBackoff?
Like, for example, let's say the node resource fit plugin rejected the Pod due to the lack of capacity and the Pod got back to unschedQ. And, we receive the NodeUpdate event that increases the Node's available capacity. 
Do you want to return QueueImmediate as well? Or do you want to return QueueAfterBackoff in that case, while wanting to return QueueImmediately here in NodeAffinity?

I'd like to clarify/make an agreement on the difference between when to return QueueImmediate and when to return QueueAfterBackoff.

Reply:
I think there are three cases to handle the event for each plugin:
1. The event doesn't help at all, we return skip.
2. The event can help, but the plugin can't be sure after the event, the pod will pass its filter. Return `QueueAfterBackoff`.
For example:  In `interpodaffinity`, a pod which conflicts with this pod is deleted, it surely helps, but the pod may still fail `interpodaffinity's` filter again.
3. The event can help, and the plugin can make sure the pod will pass its filter after this event, the plugin should return `QueueImmediately`.
For example: The node resource fit plugin rejected the Pod due to the lack of capacity. And, we receive the NodeUpdate event that increases the Node's available capacity to enough for the pod.

Overall, my thought is, when one plugin is making the decision, it shouldn't consider other plugins, it should treat this like there is only one plugin enabled.

And to avoid a unschedulabe pod got rescheduled again and again, we can consider lower the priority of `QueueImmediately` in queue side.

Reply:
-1 
That seems too aggressive because we cannot avoid the situation that high priority Pods go to activeQ again and again while blocking lower priority Pods, even if we change how we handle QueueImmediately. (Like the case that high-priority Pods are rejected by **only one** plugin while there are many lower priority Pods in the queue.) 

Also, 
> The event can help, and the plugin can make sure the pod will pass its filter after this event, the plugin should return QueueImmediately.

this is not 100% true because the cluster's situation keep changing every moment. If 100 Pods are rejected by the lack of capacity, one Node gets space (NodeUpdate event), and 100 Pods are moved back to activeQ without backoff, probably only a few of them can be schedulable and other most of them will come back in the queue with `unschedulable plugin: node resource fit` again.

Reply:
Let's move this discussion in https://github.com/kubernetes/kubernetes/pull/118551#discussion_r1269193782 since it's hard to keep two similar discussions in another thread with different people.